File: server/logger.ts
/**
 * Server Logging Module
 * 
 * This module provides standardized logging functions for the server.
 * It supports different log levels and formats logs consistently.
 */

import fs from 'fs';
import path from 'path';
import { format } from 'date-fns';

// Log levels
export enum LogLevel {
  DEBUG = 'DEBUG',
  INFO = 'INFO',
  WARNING = 'WARNING',
  ERROR = 'ERROR',
}

// Configurable settings
const LOG_TO_CONSOLE = true;
const LOG_TO_FILE = true;
const LOG_LEVEL = process.env.LOG_LEVEL || LogLevel.INFO;
const LOG_DIR = path.join(process.cwd(), 'logs');

// Ensure log directory exists
if (LOG_TO_FILE) {
  try {
    if (!fs.existsSync(LOG_DIR)) {
      fs.mkdirSync(LOG_DIR, { recursive: true });
    }
  } catch (error) {
    console.error('Failed to create log directory:', error);
  }
}

/**
 * Formats a log message with timestamp, level, and contextual information
 */
function formatLogMessage(level: LogLevel, message: string, context?: Record<string, any>): string {
  const timestamp = format(new Date(), 'yyyy-MM-dd HH:mm:ss.SSS');
  const logMessage = `[${timestamp}] [${level}] ${message}`;
  
  if (context) {
    return `${logMessage} ${JSON.stringify(context)}`;
  }
  
  return logMessage;
}

/**
 * Writes a log entry to a file
 */
function writeToFile(level: LogLevel, message: string): void {
  if (!LOG_TO_FILE) return;
  
  try {
    const date = format(new Date(), 'yyyy-MM-dd');
    const logFile = path.join(LOG_DIR, `${date}.log`);
    
    fs.appendFileSync(logFile, `${message}\n`);
  } catch (error) {
    console.error('Failed to write to log file:', error);
  }
}

/**
 * Determines if a given log level should be logged based on the configured level
 */
function shouldLog(level: LogLevel): boolean {
  const levels = Object.values(LogLevel);
  const configuredIndex = levels.indexOf(LOG_LEVEL as LogLevel);
  const currentIndex = levels.indexOf(level);
  
  return currentIndex >= configuredIndex;
}

/**
 * Core logging function that handles writing to console and file
 */
function log(level: LogLevel, message: string, context?: Record<string, any>): void {
  if (!shouldLog(level)) return;
  
  const formattedMessage = formatLogMessage(level, message, context);
  
  if (LOG_TO_CONSOLE) {
    switch (level) {
      case LogLevel.DEBUG:
        console.debug(formattedMessage);
        break;
      case LogLevel.INFO:
        console.info(formattedMessage);
        break;
      case LogLevel.WARNING:
        console.warn(formattedMessage);
        break;
      case LogLevel.ERROR:
        console.error(formattedMessage);
        break;
    }
  }
  
  if (LOG_TO_FILE) {
    writeToFile(level, formattedMessage);
  }
}

/**
 * Public logging interface
 */
export const logger = {
  debug: (message: string, context?: Record<string, any>) => log(LogLevel.DEBUG, message, context),
  info: (message: string, context?: Record<string, any>) => log(LogLevel.INFO, message, context),
  warn: (message: string, context?: Record<string, any>) => log(LogLevel.WARNING, message, context),
  error: (message: string, context?: Record<string, any>) => log(LogLevel.ERROR, message, context),
};

/**
 * Express middleware for logging HTTP requests
 */
export function requestLogger(req: any, res: any, next: Function) {
  const startTime = Date.now();
  
  // Capture response finish event
  res.on('finish', () => {
    const duration = Date.now() - startTime;
    const logContext: Record<string, any> = {
      method: req.method,
      url: req.originalUrl || req.url,
      ip: req.ip || req.connection.remoteAddress,
      statusCode: res.statusCode,
      userAgent: req.get('user-agent'),
      duration: `${duration}ms`,
    };
    
    // Get authenticated user if available
    if (req.user) {
      logContext.user = req.user.id || req.user.username;
    }
    
    // Log with appropriate level based on status code
    if (res.statusCode >= 500) {
      logger.error(`HTTP ${res.statusCode}`, logContext);
    } else if (res.statusCode >= 400) {
      logger.warn(`HTTP ${res.statusCode}`, logContext);
    } else {
      logger.info(`HTTP ${res.statusCode}`, logContext);
    }
  });
  
  next();
}

// Export default logger instance
export default logger;

File: python_agents/logger.py
"""
Server Logging Module

This module provides standardized logging functions for the Python agent system.
It supports different log levels and formats logs consistently.
"""

import os
import logging
from datetime import datetime
from pathlib import Path
from typing import Dict, Any, Optional
import json


class CustomFormatter(logging.Formatter):
    """Custom formatter that adds context information to log messages"""
    
    def format(self, record):
        # Add timestamp
        timestamp = datetime.now().strftime('%Y-%m-%d %H:%M:%S.%f')[:-3]
        
        # Format base message
        log_message = f"[{timestamp}] [{record.levelname}] {record.getMessage()}"
        
        # Add context if available
        if hasattr(record, 'context') and record.context:
            log_message += f" {json.dumps(record.context)}"
        
        # Add exception info if present
        if record.exc_info:
            log_message += f"\n{self.formatException(record.exc_info)}"
        
        return log_message


class Logger:
    """Standardized logger for the agent system"""
    
    def __init__(self, name: str = "agent_system"):
        self.logger = logging.getLogger(name)
        self.logger.setLevel(logging.INFO)
        
        # Prevent duplicate handlers
        if self.logger.handlers:
            return
        
        # Console handler
        console_handler = logging.StreamHandler()
        console_handler.setFormatter(CustomFormatter())
        self.logger.addHandler(console_handler)
        
        # File handler
        log_dir = Path.cwd() / "logs"
        log_dir.mkdir(exist_ok=True)
        
        log_file = log_dir / f"{datetime.now().strftime('%Y-%m-%d')}.log"
        file_handler = logging.FileHandler(log_file)
        file_handler.setFormatter(CustomFormatter())
        self.logger.addHandler(file_handler)
    
    def debug(self, message: str, context: Optional[Dict[str, Any]] = None):
        """Log debug message"""
        extra = {'context': context} if context else {}
        self.logger.debug(message, extra=extra)
    
    def info(self, message: str, context: Optional[Dict[str, Any]] = None):
        """Log info message"""
        extra = {'context': context} if context else {}
        self.logger.info(message, extra=extra)
    
    def warn(self, message: str, context: Optional[Dict[str, Any]] = None):
        """Log warning message"""
        extra = {'context': context} if context else {}
        self.logger.warning(message, extra=extra)
    
    def error(self, message: str, context: Optional[Dict[str, Any]] = None):
        """Log error message"""
        extra = {'context': context} if context else {}
        self.logger.error(message, extra=extra)


# Create default logger instance
logger = Logger()